src/main/java/uk/gov/hmcts/reform/bulkscan/orchestrator/services/ccd/CallbackValidations.java

```diff
@@ -51,13 +51,17 @@ private CallbackValidations() {
 
     @NotNull
     static Validation<String, String> hasCaseReference(CaseDetails theCase) {
-        Object caseReference = null;
+        Object caseObj = null;
```

Comment:
I'd refactor this method to something that will perhaps be less optimal in terms of lines of code or the number of statements, but easier to read.

Reply:
I agree with @lgonczar this method is difficult to read

Reply:
yeah its just a mass of if statments ... tbh Im not sure its going to any easier to follow if I break it down.
i.e.
```java
        Object caseObj = null;
        String caseReference;
        if( theCase != null) {
            if (theCase.getData() != null) {
                caseObj = theCase.getData().get(ATTACH_TO_CASE_REFERENCE);
                if (caseObj != null) {
                    if ( caseObj instanceof String ) {
                        caseReference = ((String) caseObj).replaceAll("[^0-9]", "");
                        if (!caseReference.isEmpty()) {
                            return valid(caseReference);
                        }
                    }
                }
            }
        }
        if ( caseObj != null ){
            return invalid(format("Invalid case reference: '%s'", String.valueOf(caseObj)));
        }else{
            return invalid( "No case reference supplied");
        }
```

Reply:
I personally think the above is actually easier to follow ??

Reply:
So the Optional version of this is:
```java

    @NotNull
    static Validation<String, String> hasCaseReference(CaseDetails theCase) {
        return Optional.ofNullable(theCase)
            .map(CaseDetails::getData)
            .map(data -> data.get(ATTACH_TO_CASE_REFERENCE))
            .flatMap(CallbackValidations::getCaseReference)
            .orElseGet(() -> invalid("No case reference supplied"));
    }

    private static Optional<Validation<String, String>> getCaseReference(Object reference) {
        Optional<Validation<String, String>> refOpt = validRef(reference).map(Validation::valid);
        return refOpt.isPresent() ? refOpt : Optional.of(invalid(format("Invalid case reference: '%s'",reference)));
    }
```

Reply:
Not much better mostly because optional does not support orElseMap hence the `refOpt.isPresent() ? `
Ill quickly try *RxSingle* which is better for this.

Reply:
actually we only have v1 Rx included so I used vavr Option instead

```java
 static Validation<String, String> hasCaseReference(CaseDetails theCase) {
        return Option.of(theCase)
            .flatMap(c -> Option.of(c.getData()))
            .flatMap(data -> Option.of(data.get(ATTACH_TO_CASE_REFERENCE)))
            .flatMap(item -> Option.of(item)
                .filter(ref -> ref instanceof String)
                .map(ref -> ((String) ref))
                .map(ref -> ref.replaceAll("[^0-9]", ""))
                .filter(StringUtils::isNotEmpty)
                .map(Validation::<String, String>valid)
                .orElse(() -> Option.of(invalid(format("Invalid case reference: '%s'", item))))
            )
            .getOrElse( ()->invalid("No case reference supplied"));
    }
```
I think I prefer this version maybe extract the flatmap though.

Reply:
The reason is that its input data ... every layer can be null so we have to check.
Otherwise the code is not *solid* and will break.

Reply:
Moved this back to using Optional since vavr might be to big a step and with java 9+ would look a little simpler.